doc/adr/0021-undo-redo.md

```diff
@@ -0,0 +1,101 @@
+
+This is the list of undo operations for each atomic:
+  - undo operation map
+      - block
+          - :block/new
+              - undo is :block/remove for uid
+          - :block/save
+              - undo is :block/save to previous str
+          - :block/open
+              - undo is :block/open with inverse bool
+          - :block/remove
+              - undo is
+                  - paste of previous IR for block uid in previous position
+                  - :block/save for all the ref in str that were replaced in remove
+          - :block/move
+              - undo is move to previous position
+      - page
+          - :page/new
+              - undo is :page/remove for title
+          - :page/rename
+              - undo is :page/rename to previous title
+          - :page/remove
+              - undo is
+                  - is paste of previous IR for title
+                  - :block/save for every ref stripped by remove
+                  - :shortcut/new + shortcut/move to previous position if any
+          - :page/merge
+              - undo is 
+                  - :page/new to previous title
+                  - :shortcut/new + shortcut/move to previous position if any
+                  - :block/move to all moved blocks
+                  - :block/save to previous string for all blocks affected by rename
+      - shortcut
+          - :shortcut/new
+              - undo is :shortcut/remove
+          - :shortcut/remove
+              - undo is
+                  - :shortcut/new 
+                  - :shortcut/move to previous position
+          - :shortcut/move
+              - undo is shortcut/move to previous position
+
+Each client will hold the last X operations they performed, and the database state they were performed over. The undo operation is computed only when undo is triggered, not ahead of time. The saved db state should be updated each time the operation order changes and the optimistic state suffers a rollback.
+
+Since undo operations can trivially generate much larger than the original, we must pay special attention to payload limits in the client and server. We know some of these limits right now but do not enforce them, so we need to start enforcing them for undo/redo to work reliably.
+
+Similarly to limits, undo also puts extra stress on the resolution error cases since it will trivially generate uncommon situations. We must make sure that invariants are enforced on resolution. Examples include: cannot rename to an existing page, cannot rename a page that does not exist, etc.
+
+A few of the undo operations rely on a view of the current data via IR (internal representation), which is then converted to a set of operations that create that structure. Undo again adds extra stress to this functionality via enhanced usage. We know it's possible to end up with unlinked refs from the order of block creation. We can address that issue by creating all blocks before adding their content.
+
```

Comment:
I'm missing description of `Redo`, which relies on being able to resolve `Undo` for any composite.
And the importance of `Undo` + `Redo` has to be last true state (just like Figma).

Reply:
Expanded the redo description back at the start of `Approach` into its own paragraph.